jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -651,83 +684,83 @@ public String toString()
+                            job = _jobs.take();
+                        else
                         {
-                            if (_idleTimeout <= 0)
-                                job = _jobs.take();
-                            else
+                            // maybe we should shrink?
+                            int size = _threadsStarted.get();
+                            if (size > _minThreads)
                             {
-                                // maybe we should shrink?
-                                final int size = _threadsStarted.get();
-                                if (size > _minThreads)
+                                long last = _lastShrink.get();
+                                long now = System.nanoTime();
+                                if (last == 0 || (now - last) > TimeUnit.MILLISECONDS.toNanos(_idleTimeout))
                                 {
-                                    long last = _lastShrink.get();
-                                    long now = System.nanoTime();
-                                    if (last == 0 || (now - last) > TimeUnit.MILLISECONDS.toNanos(_idleTimeout))
+                                    if (_lastShrink.compareAndSet(last, now))
                                     {
-                                        if (_lastShrink.compareAndSet(last, now) && _threadsStarted.compareAndSet(size, size - 1))
-                                        {
-                                            shrink = true;
-                                            break loop;
-                                        }
+                                        _threadsStarted.decrementAndGet();
+                                        shrink = true;
+                                        break loop;
                                     }
                                 }
-                                job = idleJobPoll();
                             }
+
+                            job = _jobs.poll(_idleTimeout, TimeUnit.MILLISECONDS);
                         }
                     }
-                    finally
+
+                    // run job
+                    if (job != null)
                     {
-                        if (_threadsIdle.decrementAndGet() == 0)
+                        if (idle)
                         {
-                            startThreads(1);
+                            idle = false;
+                            if (_threadsIdle.decrementAndGet() == 0)
+                                startThreads(1);
+                        }
+
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("run {}", job);
+                        runJob(job);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("ran {}", job);
+                        if (Thread.interrupted())
+                        {
+                            ignore = true;
+                            break loop;
```

Comment:
This logic was there before, but I think it's overdoing things.

If the thread was waiting, interrupting it causes `InterruptedException` which we handle below, and we exit the thread.
If the thread was not waiting, interrupting it is just flipping the interrupt flag; if the thread does not wait after the interrupt, it will complete the job and here we correctly reset the interrupt flag by calling `Thread.interrupted()`, so this thread is now good to be reused: it completed the job without exceptions, and we have reset the interrupt flag. I don't see a reason to exit the thread. 

Going back in history, apparently we added this for https://bugs.eclipse.org/bugs/show_bug.cgi?id=416477 in 16d22828fd96c2970c8c4f9292df3c175bea95c7, but there is no explanation why this addition would fix the issue (which seems unrelated). @gregw do you recall?

Had not `Thread.interrupted()` reset the interrupt flag, I would have agreed that the thread was unusable and that we had to exit it. 
But since the interrupt flag is reset, I think we can reuse it.

Reply:
I think this is here because doStop interrupts threads to get them to exit... but I guess so long as we check isRunning we can reuse the thread.